processor/src/main/java/org/mapstruct/ap/internal/model/NestedTargetPropertyMappingHolder.java

```diff
@@ -0,0 +1,212 @@
+                }
+                handledTargets.add( entryByTP.getKey().getName() );
+            }
+
+            return new NestedTargetPropertyMappingHolder( processedSourceParameters, handledTargets, propertyMappings );
+        }
+
+        /**
+         * The target references are popped. The {@code List<}{@link Mapping}{@code >} are keyed on the unique first
+         * entries of the target references.
+         *
+         * So, take
+         *
+         * targetReference 1: propertyEntryX.propertyEntryX1.propertyEntryX1a
+         * targetReference 2: propertyEntryX.propertyEntryX2
+         * targetReference 3: propertyEntryY.propertyY1
+         * targetReference 4: propertyEntryZ
+         *
+         * will be popped and grouped into entries:
+         *
+         * propertyEntryX - List ( targetReference1: propertyEntryX1.propertyEntryX1a,
+         * targetReference2: propertyEntryX2 )
+         * propertyEntryY - List ( targetReference1: propertyEntryY1 )
+         *
+         * The key will be the former top level property, the MappingOptions will contain the remainders.
+         *
+         * So, 2 cloned new MappingOptions with popped targetReferences. Also Note that the not nested targetReference4
+         * disappeared.
+         *
+         * @return See above
+         */
+        public Map<PropertyEntry, List<Mapping>> groupByPoppedTargetReferences(MappingOptions mappingOptions) {
+            Map<String, List<Mapping>> mappings = mappingOptions.getMappings();
+            // group all mappings based on the top level name before popping
+            Map<PropertyEntry, List<Mapping>> mappingsKeyedByProperty = new HashMap<PropertyEntry, List<Mapping>>();
+            for ( List<Mapping> mapping : mappings.values() ) {
+                Mapping newMapping = first( mapping ).popTargetReference();
+                if ( newMapping != null ) {
+                    // group properties on current name.
+                    PropertyEntry property = first( first( mapping ).getTargetReference().getPropertyEntries() );
+                    if ( !mappingsKeyedByProperty.containsKey( property ) ) {
+                        mappingsKeyedByProperty.put( property, new ArrayList<Mapping>() );
+                    }
+                    mappingsKeyedByProperty.get( property ).add( newMapping );
+                }
+            }
+
+            return mappingsKeyedByProperty;
+        }
+
+        /**
+         * Splits the MappingOptions into possibly more MappingOptions based on each source method parameter type.
+         *
+         * Note: this method is used for forging nested update methods. For that purpose, the same method with all
+         * joined mappings should be generated. See also: NestedTargetPropertiesTest#shouldMapNestedComposedTarget
+         *
+         * @return the split mapping options.
+         */
+        public Map<Parameter, MappingOptions> groupBySourceParameter(List<Mapping> mappings) {
```

Comment:
You are right, I think that it would be better. Just to return `Map<Parameter, List<Mapping>>` and then we just create the `MappingOptions` before we create the `PropertyMapping`